Change default ENI property FirstInterfaceIndex to 0 and improve IPAM logic in ENI & Azure modes

christarazi opened the pull request:
See commit msgs.

Fixes: https://github.com/cilium/cilium/pull/13865
Fixes: https://github.com/cilium/cilium/issues/13217

christarazi committed:
defaults, eni: Change FirstInterfaceIndex to 0

The motivation to change the default is to avoid issues with IPAM issues
when Cilium is running on smaller instances types. Given that most of
Cilium users would be running on these instances types, the default
should be in their favor. For more advanced users, they then should edit
the CiliumNode resource to set the FirstInterfaceIndex to whatever suits
their needs.

This commit also modifies the unit tests where appropriate to assert on
the new default. A few tests were left with FirstInterfaceIndex equal
to 1 for coverage.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
ipam, nodediscovery: Correct logic for PreAllocate

This commit effectively reverts f0f948d781 ("ipam: Fix ENI IPAM on
smaller instance types") as the majority of the logic it implemented is
no longer needed, since the default for FirstInterfaceIndex has been
changed to 0. While we're at it, correct the logic for computing the
PreAllocate value which is vital for ENI (and Azure) mode.

This should allow all instance types to be used with Cilium, including
the smallest ones.

Fixes: f0f948d781 ("ipam: Fix ENI IPAM on smaller instance types")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
ipam: Remove delay in refreshing IPAM nodes

There was an artificial delay to refresh the IPAM nodes (ENI & Azure) that
could cause Cilium to hang until the controller runs.[1]

The NodeManager is kicked-off by the Cilium Operator by calling
(*NodeManager).Start(). Inside here, said controller calls
(*NodeManager).Resync() which iterates over a list of nodes that need IP
addresses allocated. This eventually triggers the trigger
"ipam-node-k8s-sync-x" where "x" is the node name. This triggers
ipam.(*Node).syncToAPIServer() where the finalizing of the
initialization of the CiliumNode resource takes place. At that point,
the PreAllocate value is in-place and IP allocation can occur. The
resource is updated with the value and pushed to the apiserver. The
agent then picks up the change and can begin allocating IPs to pods and
so on. If Cilium Operator is delayed in doing this, the agent sits in a
loop, logging a msg.[1]

This commit removes this potential delay that was hardcoded to one
minute.

[1]:

```
log.WithFields(logFields).WithField(
	logfields.HelpMessage,
	"Check if cilium-operator pod is running and does not have any warnings or error messages.",
).Info("Waiting for IPs to become available in CRD-backed allocation pool")
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/ipam/node.go:
```diff
@@ -203,6 +200,7 @@ func (n *Node) getPreAllocate() int {
 	if n.resource.Spec.ENI.PreAllocate != 0 {
 		return n.resource.Spec.ENI.PreAllocate
 	}
+
```
Nit: unnecessary change

tgraf commented on pkg/aws/eni/node.go:
```diff
@@ -538,3 +539,43 @@ func (n *Node) GetMaximumAllocatableIPv4() int {
+		return minimum
+	}
+
+	return math.IntMin(minimum, limits.Adapters*limits.IPv4)
```
I guess this should be something like this in case `FirstInterfaceIndex` is not `0`.

``` go
if index >= limits.Adapters {
   return 0
}

return math.IntMin(minimum, (limits.Adapters-index)*limits.IPv4)
```

christarazi replied:
Ah good catch. Fixed. 

Yep will add an entry to the upgrade guide.

tgraf reviewed (changes requested):
LGTM

Considering the migration. Existing nodes will continue with index 1. New nodes will start with 0. I think this warrants an entry in the upgrade guide to ensure users understand that the index must be set to 1 if the current behavior should be kept.

christarazi committed:
docs: first-interface-index new ENI default

Mention that new nodes will default to 0 and existing nodes will
continue to use 1 (first-interface-index).

Signed-off-by: Chris Tarazi <chris@isovalent.com>

qmonnet commented on pkg/aws/eni/node.go:
```diff
@@ -538,3 +539,49 @@ func (n *Node) GetMaximumAllocatableIPv4() int {
+// must be allocated to the instance.
+func (n *Node) GetMinimumAllocatableIPv4() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
```
My knowledge of mutexes in Golang is limited, so this is probably a naive question. Isn't the `n.mutex.RLock()` above defeated by the `n.mutex.Lock()` that you take before calling the `GetMinimumAllocatableIPv4()` function?

christarazi replied:
Good question. Looking at the code, I see why you're confused. The answer doesn't have to do with anything specific to Golang mutexes. The `Node` in `GetMinimumAllocatableIPv4()` refers to the `pkg/aws/eni.Node` type, whereas at the site of the caller where the `Lock()` takes place as you observed is of type `pkg/ipam.Node`. These mutexes are locking different structures, although they are named the same, so it can be confusing to misleading to readers.  :)

qmonnet replied:
Ah I see, ok thank you! :)

qmonnet reviewed (commented):
One question below to better understand what you do with the mutexes, but looks good to me otherwise

Robert-Stam commented on Documentation/operations/upgrade.rst:
```diff
@@ -347,6 +347,10 @@ Annotations:
+* For AWS ENI IPAM mode, Cilium has changed the ``first-interface-index``
+  default from ``1`` to ``0``. Existing nodes in the cluster will continue to
+  use ``1`` and new nodes will use ``0``. In order to stay with Cilium's
+  current behavior, set the value to ``0`` in the CiliumNode resource.
```
``0`` --> ``1``

christarazi replied:
Good catch, thanks. Fixed

Robert-Stam reviewed (commented):
1 small comment issue, further LGTM, however see: https://cilium.slack.com/archives/C53TG4J4R/p1612348650081200?thread_ts=1610618291.011600&cid=C53TG4J4R

qmonnet reviewed (approved):
Looks good from my side.

tgraf commented on pkg/aws/eni/node.go:
```diff
@@ -538,3 +539,49 @@ func (n *Node) GetMaximumAllocatableIPv4() int {
+		return minimum
+	}
+
+	index := *n.k8sObj.Spec.ENI.FirstInterfaceIndex
```
Is this safe? Can't `FirstInterfaceIndex` be nil?

christarazi replied:
Yes, there is a check for it already a few lines above.

tgraf reviewed (changes requested):
Please update the upgrade instructions a bit more to make it clear what the implications of the change are:


> For AWS ENI IPAM mode, Cilium has changed the ``first-interface-index``
  default from ``1`` to ``0``. This means that pods will start using IPs of ``eth0`` instead of ``eth1``. This allows using the maximum number of IPs available on an instance by default. **Be aware:** Depending on your security groups configuration of the ``eth0`` interface, pods may be associated with a different security group all of a sudden. In order to stay with Cilium's current behavior, set the value to ``1`` in the CiliumNode resource.
